File: BreastCancer.R

#Ismail Arda Tuna
#240201031

#Libraries that are used

library(stringr)
library(ggplot2)
library(class)
library(Metrics)
library(caret)

#Normalize Data Function

normalizeData <- function(x) {
  num <- x - min(x)
  denom <- max(x) - min(x)
  return (num/denom)
}

#Remove Outliers Function

omitOutliers <- function(allcols,a_col){
  quartiles <- quantile(a_col, probs=c(.25, .75), na.rm = FALSE)
  iqr <- IQR(a_col)
  upper_fence <-  quartiles[2]+1.5*iqr 
  lower_fence<- quartiles[1]-1.5*iqr 
  vector_without_outliers<- subset(allcols, a_col > (lower_fence) & a_col < (upper_fence))
  return (vector_without_outliers)  
}

#Read File

readFile <- read.csv("wdbc_edit.data",TRUE,",")
class(readFile)
str(readFile)
dim(readFile)
head(readFile)
column <- data.frame(readFile)
dataset <- column[3:32]

summary(column$radius_mean)
IQR(column$radius_mean)
#Which visualization techniques would be more appropriate for analyzing this data (breast cancer)?

#Box Plot

boxplot(dataset[,2], main = "Texture Mean With Outliers", xlab="Texture Mean", ylab="Values")

#Histogram Plot

hist(dataset[,5], main = "Smoothness Mean With Outliers", xlab="Smoothness Mean", ylab="Values")

# Scatter Plot 

plot(dataset[,25], main = "Smoothness Worst Mean With Outliers", ylab="Smoothness Worst Mean", xlab="Values")

#Outlier Examples with 3 Visualization Techniques

# Box Plot 1 

dataset_without_outliers <- omitOutliers(dataset,dataset[,1])
column_without_outliers <- dataset_without_outliers[,1]
boxplot(dataset[,1], main = "Radius Mean With Outliers", xlab="Radius Mean", ylab="Values")
boxplot(column_without_outliers,  main = "Radius Mean Without Outliers", xlab="Radius Mean", ylab="Values")

# Box Plot 2 

dataset_without_outliers <- omitOutliers(dataset,dataset[,4])
column_without_outliers <- dataset_without_outliers[,4]
boxplot(dataset[,4], main = "Area Mean With Outliers", xlab="Area Mean", ylab="Values")
boxplot(column_without_outliers,  main = "Area Mean Without Outliers", xlab="Area Mean", ylab="Values")

# Box Plot 3 

dataset_without_outliers <- omitOutliers(dataset,dataset[,7])
column_without_outliers <- dataset_without_outliers[,7]
boxplot(dataset[,7], main = "Concavity Mean With Outliers", xlab="Concavity Mean", ylab="Values")
boxplot(column_without_outliers,  main = "Concavity Mean Without Outliers", xlab="Concavity Mean", ylab="Values")

# Hist Plot 1

dataset_without_outliers <- omitOutliers(dataset,dataset[,10])
column_without_outliers <- dataset_without_outliers[,10]
hist(dataset[,10], main = "Fractal Dimension Mean With Outliers", xlab="Fractal Dimension", ylab="Values")
hist(column_without_outliers,  main = "Fractal Dimension Mean Without Outliers", xlab="Fractal Dimension", ylab="Values")

# Hist Plot 2

dataset_without_outliers <- omitOutliers(dataset,dataset[,13])
column_without_outliers <- dataset_without_outliers[,13]
hist(dataset[,13], main = "Perimeter Standart Error With Outliers", xlab="Perimeter Standart Error", ylab="Values")
hist(column_without_outliers,  main = "Perimeter Standart Error Without Outliers", xlab="Perimeter Standart Error", ylab="Values")

# Hist Plot 3

dataset_without_outliers <- omitOutliers(dataset,dataset[,16])
column_without_outliers <- dataset_without_outliers[,16]
hist(dataset[,16], main = "Compactness Standart Error With Outliers", xlab="Compactness Standart Error", ylab="Values")
hist(column_without_outliers,  main = "Compactness Standart Error Without Outliers", xlab="Compactness Standart Error", ylab="Values")

# Scatter Plot 1

dataset_without_outliers <- omitOutliers(dataset,dataset[,19])
column_without_outliers <- dataset_without_outliers[,19]
plot(dataset[,19], main = "Symmetry Standart Error With Outliers", ylab="Symmetry Standart Error", xlab="Values")
plot(column_without_outliers,  main = "Symmetry Standart Error Without Outliers", ylab="Symmetry Standart Error", xlab="Values")

# Scatter Plot 2

dataset_without_outliers <- omitOutliers(dataset,dataset[,22])
column_without_outliers <- dataset_without_outliers[,22]
plot(dataset[,22], main = "Texture Worst Mean With Outliers", ylab="Texture Worst Mean", xlab="Values")
plot(column_without_outliers,  main = "Texture Worst Mean Without Outliers", ylab="Texture Worst Mean", xlab="Values")

# Scatter Plot 3

dataset_without_outliers <- omitOutliers(dataset,dataset[,25])
column_without_outliers <- dataset_without_outliers[,25]
plot(dataset[,25], main = "Smoothness Worst Mean With Outliers", ylab="Smoothness Worst Mean", xlab="Values")
plot(column_without_outliers,  main = "Smoothness Worst Mean Without Outliers", ylab="Smoothness Worst Mean", xlab="Values")

#PART 4 #Pre-processing(Normalize Data)

x <- dataset
y <- column$diagnosis_malignant_benign
y_labels <- levels(y)
x_norm <- as.data.frame(lapply(x, normalizeData))

# Check the normalization

boxplot(x, main="Initial Data")
boxplot(x_norm, main="Normalized Data")


#PART 5 #Data Mining

# set the seed to make reproducible

set.seed(123)
train_ind <- sample(seq_len(nrow(x_norm)), size = floor(0.75 * nrow(x_norm)))
x_train = x_norm[train_ind,]
x_test = x_norm[-train_ind,]
y_train = y[train_ind]
y_test = y[-train_ind]

# Check data size

dim(x_train)
dim(x_test)

#KNN Part

data_pred <- knn(train = x_train, test = x_test, cl = y_train, k=5)
data_pred
accuracy(data_pred, y_test)

# Confusion matrix

table(data_pred, y_test)

# Confusion matrix with overall statistics

confusionMatrix(data_pred, y_test)

#PART 6(Language Comparison)
boxplot(dataset[,1], main = "Radius Mean With Outliers", xlab="Radius Mean", ylab="Values")

File: BreastCancer.py
#Ismail Arda Tuna
#240201031
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn import neighbors
from sklearn.metrics import accuracy_score, confusion_matrix

#The function below is retrieved from 
#https://towardsdatascience.com/ways-to-detect-and-remove-the-outliers-404d16608dba

def omitOutliers (col):
    q1 = col.quantile(0.25)
    q3 = col.quantile(0.75)
    iqr = q3 - q1
    upper_fence = q3 + 1.5*iqr
    lower_fence = q1 - 1.5*iqr
    col_without_outlier = col[((~(col < lower_fence)) & (~(col > upper_fence)))]
    return col_without_outlier
        
read_file = pd.read_csv("wdbc_edit.data")
dataset = read_file[read_file.columns[2:32]]
target = read_file[read_file.columns[1]]
x = read_file[read_file.columns[2]].to_frame()


#Which visualization techniques would be more appropriate for analyzing this data (breast cancer)?

#Box Plot
print("-----------------------------------------------------------")
print("          Visualization Techniques Examples")
print("")
print("                    BOX PLOT")
print("                    --------")
plt.xlabel("Concavity Mean")
plt.ylabel("Value")
plt.title("Concavity Mean With Outliers")
plt.boxplot(dataset[dataset.columns[6]])
plt.show()
#Histogram Plot

print("")
print("                 HISTOGRAM PLOT")
print("                 --------------")
plt.xlabel("Area Mean")
plt.ylabel("Value")
plt.title("Area Mean With Outliers")
plt.hist(dataset[dataset.columns[3]])
plt.show()

# Scatter Plot

print("")
print("                 SCATTER PLOT") 
print("                 ------------")
plt.ylabel("Radius Mean")
plt.xlabel("Value")
plt.title("Radius Mean With Outliers")
plt.scatter(dataset[dataset.columns[0]].index,dataset[dataset.columns[0]])
plt.show()

#Outlier Examples with 3 Visualization Techniques

# Box Plot  
print("-----------------------------------------------------------")
print("          Outlier Handling Examples")
print("")
print("                    BOX PLOT")
print("                    --------")
plt.xlabel("Compactness Standart Error")
plt.ylabel("Value")
plt.title("Compactness Standart Error With Outliers")
plt.boxplot(dataset[dataset.columns[15]])
plt.show()

plt.xlabel("Compactness Standart Error")
plt.ylabel("Value")
plt.title("Compactness Standart Error Without Outliers")
plt.boxplot(omitOutliers(dataset[dataset.columns[15]]))
plt.show()

# Hist Plot 

print("")
print("                 HISTOGRAM PLOT")
print("                 --------------")
plt.xlabel("Perimeter Standart Error")
plt.ylabel("Value")
plt.title("Perimeter Standart Error With Outliers")
plt.hist(dataset[dataset.columns[12]])
plt.show()

plt.xlabel("Perimeter Standart Error")
plt.ylabel("Value")
plt.title("Perimeter Standart Error Without Outliers")
plt.hist(omitOutliers(dataset[dataset.columns[12]]))
plt.show()
# Scatter Plot 

print("")
print("                 SCATTER PLOT") 
print("                 ------------")
plt.ylabel("Symmetry Standart Error")
plt.xlabel("Value")
plt.title("Symmetry Standart Error With Outliers")
plt.scatter(dataset[dataset.columns[18]].index,dataset[dataset.columns[18]])
plt.show()

plt.ylabel("Symmetry Standart Error")
plt.xlabel("Value")
plt.title("Symmetry Standart Error Without Outliers")
plt.scatter(omitOutliers(dataset[dataset.columns[18]]).index,omitOutliers(dataset[dataset.columns[18]]))
plt.show()



##PART 4 #Pre-processing

x = dataset
y = target
print("")
print(y.head(4))
print("")
print(y.tail(4))

#(Normalize Data)

min_max_scaler = MinMaxScaler()
x_norm = min_max_scaler.fit_transform(x)

# Check the normalization 

print("")
plt.title("Initial Data")
plt.boxplot(x)
plt.show()
print("")
plt.title("Normalized Data")
plt.boxplot(x_norm)
plt.show()

#PART 5 #Data Mining

# set the random state to make reproducible

x_train,x_test,y_train,y_test=train_test_split(x, y, test_size = .25 , random_state = 123 )

# Check data size

print("")
print ("Train size" + " ------------>  " + str(x_train.shape))
print("")
print ("Test size" + "  ------------>  " + str(x_test.shape))

#KNN Part

classifier=neighbors.KNeighborsClassifier( n_neighbors = 5 )
classifier.fit(x_train,y_train)
data_pred = classifier.predict(x_test)

#Accuracy

print("")
print ("Accuracy" + "  ------------>  "+str(accuracy_score(y_test, data_pred)))

#Confusion Matrix

print("")
print("Confusion Matrix:")
print("")
print (confusion_matrix(y_test, data_pred))


#PART 6 (Language Comparison)

#Box Plot

box_sample = dataset[dataset.columns[0]]
plt.xlabel("Radius Mean")
plt.ylabel("Value")
plt.title("Radius Mean With Outliers")
plt.boxplot(box_sample)
plt.show()
